src/Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv/Internal/ListenerContext.cs

```diff
@@ -28,18 +28,65 @@ protected UvStreamHandle CreateAcceptSocket()
             switch (EndPointInformation.Type)
             {
                 case ListenType.IPEndPoint:
-                case ListenType.FileHandle:
-                    var tcpHandle = new UvTcpHandle(TransportContext.Log);
-                    tcpHandle.Init(Thread.Loop, Thread.QueueCloseHandle);
-                    tcpHandle.NoDelay(EndPointInformation.NoDelay);
-                    return tcpHandle;
+                    return AcceptTcp();
                 case ListenType.SocketPath:
-                    var pipeHandle = new UvPipeHandle(TransportContext.Log);
-                    pipeHandle.Init(Thread.Loop, Thread.QueueCloseHandle);
-                    return pipeHandle;
+                    return AcceptPipe();
+                case ListenType.FileHandle:
+                    return AcceptHandle();
                 default:
                     throw new InvalidOperationException();
             }
         }
+
+        private UvTcpHandle AcceptTcp()
+        {
+            var socket = new UvTcpHandle(TransportContext.Log);
+
+            try
+            {
+                socket.Init(Thread.Loop, Thread.QueueCloseHandle);
+                socket.NoDelay(EndPointInformation.NoDelay);
+            }
+            catch
+            {
+                socket.Dispose();
+                throw;
+            }
+
+            return socket;
+        }
+
+        private UvPipeHandle AcceptPipe()
+        {
+            var pipe = new UvPipeHandle(TransportContext.Log);
+
+            try
+            {
+                pipe.Init(Thread.Loop, Thread.QueueCloseHandle);
+            }
+            catch
+            {
+                pipe.Dispose();
+                throw;
+            }
+
+            return pipe;
+        }
+
+        private UvStreamHandle AcceptHandle()
+        {
+            switch (EndPointInformation.HandleType)
+            {
+                case FileHandleType.Auto:
+                    //  Handle type detection must be done by listener, refreshing endpoint information
+                    throw new InvalidOperationException();
```

Comment:
Might as well throw the comment in the exception message in case this somehow throws someday.